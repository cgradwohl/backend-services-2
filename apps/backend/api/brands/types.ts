import { ApiRequestContext } from "~/lib/lambda-response";
import {
  IApiBrandsGetResponse,
  IApiBrandsListResponse,
  IApiBrandsPostRequest,
  IApiBrandsPostResponse,
  IApiBrandsPutRequest,
  IApiBrandsPutResponse,
} from "~/types.public";

export type ApiBrandsResponse =
  | IApiBrandsGetResponse
  | IApiBrandsListResponse
  | IApiBrandsPostResponse
  | IApiBrandsPutResponse;

export type GetFn = (
  context: ApiRequestContext
) => Promise<{
  body: IApiBrandsGetResponse;
}>;

export type ListFn = (
  context: ApiRequestContext
) => Promise<{
  body: IApiBrandsListResponse;
}>;

export type PostRequestBody = IApiBrandsPostRequest;

export type PostFn = (
  context: ApiRequestContext
) => Promise<{
  body: IApiBrandsPostResponse;
}>;

export type PutRequestBody = IApiBrandsPutRequest;

export type PutFn = (
  context: ApiRequestContext
) => Promise<{
  body: IApiBrandsPutResponse;
}>;

export type RemoveFn = (
  context: ApiRequestContext
) => Promise<{
  status: 204;
}>;
