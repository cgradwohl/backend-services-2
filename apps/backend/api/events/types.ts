import { ApiRequestContext } from "~/lib/lambda-response";
import {
  IApiEventsGetResponse,
  IApiEventsListResponse,
  IApiEventsPutRequest,
  IApiEventsPutResponse,
} from "~/types.public";

export type ApiEventsResponse = IApiEventsGetResponse | IApiEventsListResponse;

export type GetFn = (
  context: ApiRequestContext
) => Promise<{
  body: IApiEventsGetResponse;
}>;

export type ListFn = (
  context: ApiRequestContext
) => Promise<{
  body: IApiEventsListResponse;
}>;

export type PutRequestBody = IApiEventsPutRequest;

export type PutFn = (
  context: ApiRequestContext
) => Promise<{
  body: IApiEventsPutResponse;
}>;

export type RemoveFn = (
  context: ApiRequestContext
) => Promise<{
  status: 204;
}>;
