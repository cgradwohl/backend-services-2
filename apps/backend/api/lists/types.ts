import { ApiRequestContext, Response } from "~/lib/lambda-response";

import {
  IApiDeleteListItemResponse,
  IApiGetListItemResponse,
  IApiGetListItemsResponse,
  IApiGetListItemSubscriptionsResponse,
  IApiListItemSubscribeResponse,
  IApiListItemUnsubscribeResponse,
  IApiPutListItemResponse,
  IApiRestoreListItemResponse,
} from "~/types.public";

export type HttpEventHandler<T extends Response<any>> = (
  context: ApiRequestContext
) => Promise<T>;

export type ListApiResponse =
  | IApiDeleteListItemResponse
  | IApiGetListItemResponse
  | IApiGetListItemsResponse
  | IApiGetListItemSubscriptionsResponse
  | IApiListItemSubscribeResponse
  | IApiListItemUnsubscribeResponse
  | IApiPutListItemResponse
  | IApiRestoreListItemResponse;
