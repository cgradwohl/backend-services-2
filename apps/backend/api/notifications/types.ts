import { ApiRequestContext } from "~/lib/lambda-response";
import {
  IApiNotificationGetResponse,
  IApiNotificationGetSubmissionChecksResponse,
  IApiNotificationListResponse,
  IApiNotificationPutSubmissionChecksResponse,
} from "~/types.public";

export type ApiNotificationResponse =
  | IApiNotificationGetResponse
  | IApiNotificationListResponse
  | IApiNotificationGetSubmissionChecksResponse
  | IApiNotificationPutSubmissionChecksResponse;

export type GetFn = (
  context: ApiRequestContext
) => Promise<{
  body: IApiNotificationGetResponse;
}>;

export type PostFn = (
  context: ApiRequestContext
) => Promise<{
  status: number;
}>;

export type PutFn = (
  context: ApiRequestContext
) => Promise<{
  status: number;
}>;

export type ListFn = (
  context: ApiRequestContext
) => Promise<{
  body: IApiNotificationListResponse;
}>;

export type GetSubmissionChecksFn = (
  context: ApiRequestContext
) => Promise<{
  body: IApiNotificationGetSubmissionChecksResponse;
}>;

export type PutSubmissionChecksFn = (
  context: ApiRequestContext
) => Promise<{
  body: IApiNotificationPutSubmissionChecksResponse;
}>;

export type DeleteSubmissionChecksFn = (
  context: ApiRequestContext
) => Promise<{
  status: 204 | 409;
}>;
