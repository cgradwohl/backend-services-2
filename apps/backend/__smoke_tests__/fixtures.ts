const _emailEvents = (() => {
  if (process.env.SMOKE_TEST_EVENT) {
    return [process.env.SMOKE_TEST_EVENT];
  }

  if (process.env.STAGE === "staging") {
    return ["SAPBJ1JWR44V2MQ3NBRFRVAC594W"];
  }
  if (process.env.STAGE === "production") {
    return ["14171CNHXG4JHXNH2NRMBYJ0Y20P"];
  }

  return [];
})();

export const emailEvents = _emailEvents;
