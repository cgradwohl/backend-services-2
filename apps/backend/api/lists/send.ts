import { nanoid } from "nanoid";

import assertStateIsValid from "~/lib/assertions/is-valid-scope-state";
import { get as getBrand, getLatest } from "~/lib/brands";
import { IBrand } from "~/lib/brands/types";
import { assertValidDataSourceConfig } from "~/lib/data-source/assert-valid-data-source-config";
import { InvalidDataSourceConfigError } from "~/lib/data-source/errors";
import enqueue from "~/lib/enqueue";
import { BadRequest, NotFound } from "~/lib/http-errors";
import { assertBody, handleIdempotentApi } from "~/lib/lambda-response";
import { assertValidPattern } from "~/lib/lists";
import { get as getList } from "~/lib/lists";
import { InvalidListSearchPatternError } from "~/lib/lists/errors";
import { IListItem } from "~/lib/lists/types";
import parseJsonObject from "~/lib/parse-json-object";
import jsonStore from "~/lib/s3";
import createTraceId from "~/lib/x-ray/create-trace-id";
import { JSONObject } from "~/types.api";
import {
  S3SendListOrPatternMessage,
  SqsSendListOrPatternMessage,
  TenantRouting,
  TenantScope,
} from "~/types.internal";
import {
  ApiSendListOrPatternRequest,
  ApiSendRequestOverride,
  ApiSendResponse,
  IProfilePreferences,
} from "~/types.public";

const enqueueSendListOrPatternMessage = enqueue<SqsSendListOrPatternMessage>(
  process.env.SQS_SEND_LIST_OR_PATTERN_QUEUE_NAME
);

const { put: putListOrPatternMessage } = jsonStore<S3SendListOrPatternMessage>(
  process.env.S3_MESSAGES_BUCKET
);

// send/list supports sending a list or a pattern that can match 1+ list(s)
// used in SendListOrPattern trigger as well
export const ListSendTypes = {
  list: "send-list",
  pattern: "send-pattern",
};

const saveAndEnqueue = async (
  messageId: string,
  tenantId: string,
  message: S3SendListOrPatternMessage,
  originalMessageId: string
) => {
  const type = message.list?.id ? ListSendTypes.list : ListSendTypes.pattern;
  const filePath = `${tenantId}/${type}-${nanoid()}.json`;
  // we use the original message id for lists to tie each event log
  // to the original list send. patterns will not appear in logs,
  // hence not worrying about standardizing on the originalMessageId
  const resolvedMessageId = originalMessageId ?? messageId;

  // save to s3
  await putListOrPatternMessage(filePath, message);

  // enqueue message
  await enqueueSendListOrPatternMessage({
    messageId,
    messageLocation: {
      path: filePath,
      type: "S3",
    },
    originalMessageId: resolvedMessageId,
    tenantId,
    type,
  });
};

const getScopedBrand = async (
  tenantId: string,
  id: string,
  state: "published" | "draft" | "submitted"
) =>
  ["published", "submitted"].includes(state) // get published brand for submitted state as well
    ? getBrand(tenantId, id, {
        extendDefaultBrand: true,
      })
    : getLatest(tenantId, id);

export const sendListOrPattern = async (
  body: ApiSendListOrPatternRequest,
  brand: IBrand,
  eventData: JSONObject,
  eventId: string,
  messageId: string,
  override: ApiSendRequestOverride,
  tenantId: string,
  list?: IListItem,
  pattern?: string,
  originalMessageId?: string,
  dryRunKey?: TenantRouting,
  scope?: TenantScope,
  eventPreferences?: IProfilePreferences
) => {
  const message: S3SendListOrPatternMessage = {
    brand,
    dataSource: body.data_source,
    dryRunKey,
    eventData,
    eventId,
    eventPreferences,
    list,
    override,
    pattern,
    scope,
  };
  await saveAndEnqueue(messageId, tenantId, message, originalMessageId);
};

export const handleListOrPatternSendRequest = async ({
  context,
  messageId,
}) => {
  const tenantId = context.tenantId;
  const dryRunKey = context.dryRunKey;
  const scope = context.scope;
  const body = assertBody<ApiSendListOrPatternRequest>(context);
  const brandId = body.brand;
  const eventId = body.event;
  const eventData = parseJsonObject(body.data);
  const eventPreferences = parseJsonObject<IProfilePreferences>(
    body.preferences
  );
  const override = parseJsonObject<ApiSendRequestOverride>(body.override);
  const pattern = body.pattern;
  const listId = body.list;
  const dataSource = body.data_source;

  if (!eventId) {
    throw new BadRequest("The 'event' parameter is required.");
  }

  if (typeof eventId !== "string") {
    throw new BadRequest("The 'event' parameter must be a string.");
  }

  if (eventData === null) {
    throw new BadRequest("The 'data' parameter must be valid JSON.");
  }

  // must have one of listId || pattern, but not both
  if (!listId && !pattern) {
    throw new BadRequest("Either 'list' or 'pattern' parameter is required.");
  }

  if (listId && pattern) {
    throw new BadRequest(
      "Only one of the following properties allowed: 'list', 'pattern'"
    );
  }

  if (pattern) {
    try {
      assertValidPattern(pattern);
    } catch (e) {
      // need to handle these validations to ensure they return as 400s
      if (e instanceof InvalidListSearchPatternError) {
        throw new BadRequest(e.message);
      }
    }
  }

  if (dataSource) {
    try {
      assertValidDataSourceConfig(dataSource);
    } catch (e) {
      // need to handle these validations to ensure they return as 400s
      if (e instanceof InvalidDataSourceConfigError) {
        throw new BadRequest(e.message);
      }
    }
  }

  const [state] = scope.split("/");
  assertStateIsValid(state);

  // valid brand required
  let brand: IBrand;
  if (brandId) {
    try {
      brand = await getScopedBrand(tenantId, brandId, state);
    } catch (err) {
      if (err instanceof NotFound) {
        throw new BadRequest(`Invalid brand (${brandId})`);
      }
      throw err;
    }

    if (!brand) {
      throw new BadRequest(`Invalid brand (${brandId})`);
    }

    if (!brand.published && ["published", "submitted"].includes(state)) {
      // brand must be published
      throw new BadRequest(`Brand (${brandId}) not published`);
    }
  }

  // valid list required
  let list: IListItem;
  if (listId) {
    try {
      list = await getList(tenantId, listId);
      if (!list?.id) {
        throw new BadRequest(`Cannot send to archived list (${listId})`);
      }
    } catch (err) {
      if (err instanceof NotFound) {
        throw new BadRequest(`Invalid list (${listId})`);
      }
      throw err;
    }
    // To prevent double sanitization
    list.id = listId;
  }

  await sendListOrPattern(
    body,
    brand,
    eventData,
    eventId,
    messageId,
    override,
    tenantId,
    list,
    pattern,
    undefined,
    dryRunKey,
    scope,
    eventPreferences
  );
};

export const handle = handleIdempotentApi<ApiSendResponse>(async (context) => {
  const messageId = createTraceId();

  await handleListOrPatternSendRequest({ context, messageId });

  return { body: { messageId } };
});
