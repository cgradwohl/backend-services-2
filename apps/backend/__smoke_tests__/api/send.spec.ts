import dotenv from "dotenv";
dotenv.config({ path: "__smoke_tests__/.env" });
import { CourierClient } from "@trycourier/courier";
import axios, { AxiosRequestConfig } from "axios";
import { RequestV2 } from "../../api/send/types";

const JEST_TIMEOUT = 60000;

const courier = CourierClient({
  authorizationToken: process.env.COURIER_AUTH_TOKEN,
  baseUrl: process.env.API_URL,
});

const smokeTestEmail =
  process.env.SMOKE_TEST_EMAIL || "trycourier@litmusemail.com";

const axiosHeaders = {
  Authorization: `Bearer ${process.env.COURIER_AUTH_TOKEN}`,
  "Content-Type": "application/json",
};

import { emailEvents } from "../fixtures";

const wait = () =>
  new Promise((resolve) => {
    setTimeout(resolve, 5000);
  });

describe("send api v1", () => {
  emailEvents.forEach((eventId) => {
    let messageId;

    test(
      `can send email notification: ${eventId}`,
      async () => {
        const response = await courier.send({
          data: {},
          eventId,
          profile: {
            email: smokeTestEmail,
          },
          recipientId: "@smoketest",
        });

        messageId = response.messageId;
        expect(messageId).toBeDefined();
      },
      JEST_TIMEOUT
    );

    test(
      `can retrieve message`,
      async () => {
        await wait();

        let attempts = 10;
        let messageStatus = await courier.getMessage(messageId);
        while (messageStatus.status === "ENQUEUED" && attempts) {
          await wait();
          messageStatus = await courier.getMessage(messageId);
          attempts = attempts - 1;
        }

        expect(
          ["SENT", "DELIVERED", "OPENED", "CLICKED"].includes(
            messageStatus.status
          )
        ).toBe(true);
      },
      JEST_TIMEOUT
    );
  });
});

describe("send api v2 - email", () => {
  let messageId;
  let messageHistory;
  let renderedEvent;

  test(
    `can send inline content`,
    async () => {
      const data: RequestV2 = {
        message: {
          content: {
            version: "2022-01-01",
            elements: [
              { type: "meta", title: "Hello, {{name}}" },
              { type: "text", content: "Good Day!" },
              {
                type: "action",
                href: "https://www.courier.com",
                align: "center",
                background_color: "#9d3789",
                style: "button",
                content: "Click Me!",
              },
            ],
          },
          data: {
            name: "Smokey",
          },
          brand: {
            version: "2022-05-17",
            colors: {
              primary: "#344563",
              secondary: "#C1B6DD",
              tertiary: "#E85178",
            },
            logo: {
              image: "https://via.placeholder.com/150",
              href: "app.courier.com",
            },
          },
          to: { email: smokeTestEmail },
          routing: {
            method: "single",
            channels: ["email"],
          },
        },
      };
      const options: AxiosRequestConfig = {
        data,
        headers: axiosHeaders,
        method: "POST",
        url: `${process.env.API_URL}/send`,
      };

      const response = await axios.request(options);
      messageId = response.data.requestId;
      console.log("messageId", messageId);
      expect(messageId).toBeDefined();
    },
    JEST_TIMEOUT
  );

  test(
    `can verify message sent`,
    async () => {
      await wait();

      let messageStatus;
      let attempts = 10;
      try {
        messageStatus = await courier.getMessage(messageId);
        while (messageStatus.status === "ENQUEUED" && attempts) {
          await wait();
          messageStatus = await courier.getMessage(messageId);
          attempts = attempts - 1;
        }
      } catch (ex) {
        console.log(`error getting message, ${messageId}:`, ex);
      }

      console.log("messageStatus", messageStatus);
      expect(
        ["SENT", "DELIVERED", "OPENED", "CLICKED"].includes(
          messageStatus.status
        )
      ).toBe(true);
    },
    JEST_TIMEOUT
  );

  test(
    `can get message history`,
    async () => {
      const options: AxiosRequestConfig = {
        headers: axiosHeaders,
        method: "GET",
        url: `${process.env.API_URL}/messages/${messageId}/history`,
      };

      let attempts = 10;
      let response;

      const doRequest = async () => {
        response = await axios.request(options);
      };

      while (!response && attempts) {
        try {
          await wait();
          await doRequest();
        } catch (ex) {
          console.log(`error getting message history, ${messageId}:`, ex);
        }

        attempts = attempts - 1;
      }

      messageHistory = response.data.results;
      expect(messageHistory.some((item) => item.type === "ENQUEUED")).toBe(
        true
      );

      renderedEvent = messageHistory.find((item) => item.type === "RENDERED");
      expect(renderedEvent).toBeTruthy();

      expect(messageHistory.some((item) => item.type === "SENT")).toBe(true);
    },
    JEST_TIMEOUT
  );

  test(
    `can get rendered html`,
    async () => {
      const hrefRegex = /<a\s+(?:[^>]*?\s+)?href=(["'])(.*?)\1/g;
      const imgSrcRegex = /<img\s+(?:[^>]*?\s+)?src=(["'])(.*?)\1/g;
      const options: AxiosRequestConfig = {
        headers: {
          ...axiosHeaders,
          "content-type": "text/html",
        },
        method: "GET",
        url: `${process.env.API_URL}${renderedEvent.output.html}`,
      };

      const response = await axios.request(options);
      const body = response.data
        .replace(hrefRegex, '<a href="REPLACED"')
        .replace(imgSrcRegex, '<img src="REPLACED"');

      expect(body).toMatchSnapshot();
    },
    JEST_TIMEOUT
  );

  test(
    `can get rendered subject`,
    async () => {
      const options: AxiosRequestConfig = {
        headers: {
          ...axiosHeaders,
          "content-type": "text/html",
        },
        method: "GET",
        url: `${process.env.API_URL}${renderedEvent.output.subject}`,
      };

      const response = await axios.request(options);
      expect(response.data).toMatchSnapshot();
    },
    JEST_TIMEOUT
  );
});
