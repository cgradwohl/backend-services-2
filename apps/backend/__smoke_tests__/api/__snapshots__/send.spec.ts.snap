// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`send api v2 - email can get rendered html 1`] = `
"<html xmlns=\\"http://www.w3.org/1999/xhtml\\"><head>
        <title>
          
        </title>
        
        
        
        
        
        <style type=\\"text/css\\">
          #outlook a { padding:0; }
          body { margin:0;padding:0;-webkit-text-size-adjust:100%;-ms-text-size-adjust:100%; }
          table, td { border-collapse:collapse;mso-table-lspace:0pt;mso-table-rspace:0pt; }
          img { border:0;height:auto;line-height:100%; outline:none;text-decoration:none;-ms-interpolation-mode:bicubic; }
          p { display:block;margin:13px 0; }
        </style>
        
        
        
        
    <style type=\\"text/css\\">
      @media only screen and (min-width:480px) {
        .mj-column-per-100 { width:100% !important; max-width: 100%; }
      }
    </style>
    
  
        <style type=\\"text/css\\">
        
        

    @media only screen and (max-width:480px) {
      table.full-width-mobile { width: 100% !important; }
      td.full-width-mobile { width: auto !important; }
    }
  
        </style>
        
        
      </head>
      <body style=\\"background-color:#f5f5f5;\\">
    
        
        
      <div style=\\"padding-bottom: 20px; background-color: #f5f5f5;\\" class=\\"c--email-body\\">
        
      
      
    
      
      <div style=\\"margin:0px auto;max-width:582px;\\">
        
        <table style=\\"width:100%;\\" role=\\"presentation\\" cellspacing=\\"0\\" cellpadding=\\"0\\" border=\\"0\\" align=\\"center\\">
          <tbody>
            <tr>
              <td style=\\"direction:ltr;font-size:0px;padding:20px 0 0 0;text-align:center;\\">
                
    
      
      <div style=\\"border-bottom: 1px solid #f5f5f5; border-top: 6px solid #344563; border-radius: 7px 7px 0 0; background: #ffffff; background-color: #ffffff; margin: 0px auto; max-width: 582px;\\" class=\\"c--email-header\\">
        
        <table style=\\"background:#ffffff;background-color:#ffffff;width:100%;\\" role=\\"presentation\\" cellspacing=\\"0\\" cellpadding=\\"0\\" border=\\"0\\" align=\\"center\\">
          <tbody>
            <tr>
              <td style=\\"border-left:1px solid #f5f5f5;border-right:1px solid #f5f5f5;direction:ltr;font-size:0px;padding:0px;text-align:center;\\">
                
            
      <div style=\\"font-size:0px;text-align:left;direction:ltr;display:inline-block;vertical-align:top;width:100%;\\" class=\\"mj-column-per-100 outlook-group-fix\\">
        
      <table width=\\"100%\\" role=\\"presentation\\" cellspacing=\\"0\\" cellpadding=\\"0\\" border=\\"0\\">
        <tbody>
          <tr>
            <td style=\\"background-color:#ffffff;vertical-align:top;padding:0px;padding-top:20px;padding-bottom:20px;padding-left:20px;\\">
              
      <table width=\\"100%\\" style=\\"\\" role=\\"presentation\\" cellspacing=\\"0\\" cellpadding=\\"0\\" border=\\"0\\">
        
            <tbody><tr>
              <td style=\\"font-size:0px;padding:0px;word-break:break-word;\\" align=\\"left\\">
                
      <table style=\\"border-collapse:collapse;border-spacing:0px;\\" role=\\"presentation\\" cellspacing=\\"0\\" cellpadding=\\"0\\" border=\\"0\\">
        <tbody>
          <tr>
            <td style=\\"width:140px;\\">
              
        <a href=\\"REPLACED\\">
          
      <img src=\\"REPLACED\\" height=\\"auto\\">
    
        </a>
      
            </td>
          </tr>
        </tbody>
      </table>
    
              </td>
            </tr>
          
      </tbody></table>
    
            </td>
          </tr>
        </tbody>
      </table>
    
      </div>
    
          
              </td>
            </tr>
          </tbody>
        </table>
        
      </div>
    
      
      
    
      
      <div style=\\"background:#ffffff;background-color:#ffffff;margin:0px auto;max-width:582px;\\">
        
        <table style=\\"background:#ffffff;background-color:#ffffff;width:100%;\\" role=\\"presentation\\" cellspacing=\\"0\\" cellpadding=\\"0\\" border=\\"0\\" align=\\"center\\">
          <tbody>
            <tr>
              <td style=\\"border-left:1px solid #f5f5f5;border-right:1px solid #f5f5f5;direction:ltr;font-size:0px;padding:8px 30px;text-align:center;\\">
                
            
      <div style=\\"font-size:0px;text-align:left;direction:ltr;display:inline-block;vertical-align:top;width:100%;\\" class=\\"mj-column-per-100 outlook-group-fix\\">
        
      <table width=\\"100%\\" role=\\"presentation\\" cellspacing=\\"0\\" cellpadding=\\"0\\" border=\\"0\\">
        <tbody>
          <tr>
            <td style=\\"background-color:#ffffff;vertical-align:top;padding:0px;\\">
              
      <table width=\\"100%\\" style=\\"\\" role=\\"presentation\\" cellspacing=\\"0\\" cellpadding=\\"0\\" border=\\"0\\">
        
      </table>
    
            </td>
          </tr>
        </tbody>
      </table>
    
      </div>
    
          
              </td>
            </tr>
          </tbody>
        </table>
        
      </div>
    
      
      
    
      
      <div style=\\"background:#ffffff;background-color:#ffffff;margin:0px auto;max-width:582px;\\" class=\\"c--block c--block-text\\">
        
        <table style=\\"background:#ffffff;background-color:#ffffff;width:100%;\\" role=\\"presentation\\" cellspacing=\\"0\\" cellpadding=\\"0\\" border=\\"0\\" align=\\"center\\">
          <tbody>
            <tr>
              <td style=\\"border-left:1px solid #f5f5f5;border-right:1px solid #f5f5f5;direction:ltr;font-size:0px;padding:8px 30px;text-align:center;\\">
                
            
      <div style=\\"font-size:0px;text-align:left;direction:ltr;display:inline-block;vertical-align:top;width:100%;\\" class=\\"mj-column-per-100 outlook-group-fix\\">
        
      <table width=\\"100%\\" role=\\"presentation\\" cellspacing=\\"0\\" cellpadding=\\"0\\" border=\\"0\\">
        <tbody>
          <tr>
            <td style=\\"background-color:transparent;vertical-align:top;padding:4px 0px;\\">
              
      <table width=\\"100%\\" style=\\"\\" role=\\"presentation\\" cellspacing=\\"0\\" cellpadding=\\"0\\" border=\\"0\\">
        
            <tbody><tr>
              <td style=\\"font-size:0px;padding:0px;word-break:break-word;\\" class=\\"c--text-text\\" align=\\"left\\">
                
      <div style=\\"font-family:Helvetica, Arial, sans-serif;font-size:14px;line-height:18px;text-align:left;color:#4C4C4C;\\">Good Day!</div>
    
              </td>
            </tr>
          
      </tbody></table>
    
            </td>
          </tr>
        </tbody>
      </table>
    
      </div>
    
          
              </td>
            </tr>
          </tbody>
        </table>
        
      </div>
    
      
      
    
      
      <div style=\\"background:#ffffff;background-color:#ffffff;margin:0px auto;max-width:582px;\\" class=\\"c--block c--block-action\\">
        
        <table style=\\"background:#ffffff;background-color:#ffffff;width:100%;\\" role=\\"presentation\\" cellspacing=\\"0\\" cellpadding=\\"0\\" border=\\"0\\" align=\\"center\\">
          <tbody>
            <tr>
              <td style=\\"border-left:1px solid #f5f5f5;border-right:1px solid #f5f5f5;direction:ltr;font-size:0px;padding:8px 30px;text-align:center;\\">
                
            
      <div style=\\"font-size:0px;text-align:left;direction:ltr;display:inline-block;vertical-align:top;width:100%;\\" class=\\"mj-column-per-100 outlook-group-fix\\">
        
      <table width=\\"100%\\" role=\\"presentation\\" cellspacing=\\"0\\" cellpadding=\\"0\\" border=\\"0\\">
        <tbody>
          <tr>
            <td style=\\"background-color:#ffffff;vertical-align:top;padding:0px;\\">
              
      <table width=\\"100%\\" style=\\"\\" role=\\"presentation\\" cellspacing=\\"0\\" cellpadding=\\"0\\" border=\\"0\\">
        
            <tbody><tr>
              <td style=\\"font-size:0px;padding:8px 0px;word-break:break-word;\\" align=\\"center\\">
                
      <table style=\\"border-collapse:separate;line-height:100%;\\" role=\\"presentation\\" cellspacing=\\"0\\" cellpadding=\\"0\\" border=\\"0\\">
        <tbody><tr>
          <td valign=\\"middle\\" style=\\"border:none;border-radius:4px;cursor:auto;mso-padding-alt:10px 20px;background:#9d3789;\\" role=\\"presentation\\" bgcolor=\\"#9d3789\\" align=\\"center\\">
            <a href=\\"REPLACED\\">
              Click Me!
            </a>
          </td>
        </tr>
      </tbody></table>
    
              </td>
            </tr>
          
      </tbody></table>
    
            </td>
          </tr>
        </tbody>
      </table>
    
      </div>
    
          
              </td>
            </tr>
          </tbody>
        </table>
        
      </div>
    
      
      
    
      
      <div style=\\"background: #ffffff; background-color: #ffffff; margin: 0px auto; max-width: 582px;\\" class=\\"c--email-footer\\">
        
        <table style=\\"background:#ffffff;background-color:#ffffff;width:100%;\\" role=\\"presentation\\" cellspacing=\\"0\\" cellpadding=\\"0\\" border=\\"0\\" align=\\"center\\">
          <tbody>
            <tr>
              <td style=\\"border-left:1px solid #f5f5f5;border-right:1px solid #f5f5f5;direction:ltr;font-size:0px;padding:0px;text-align:center;\\">
                
            
      <div style=\\"font-size:0px;text-align:left;direction:ltr;display:inline-block;vertical-align:top;width:100%;\\" class=\\"mj-column-per-100 outlook-group-fix\\">
        
      <table width=\\"100%\\" role=\\"presentation\\" cellspacing=\\"0\\" cellpadding=\\"0\\" border=\\"0\\">
        <tbody>
          <tr>
            <td style=\\"background-color:#ffffff;vertical-align:top;padding:0px;\\">
              
      <table width=\\"100%\\" style=\\"\\" role=\\"presentation\\" cellspacing=\\"0\\" cellpadding=\\"0\\" border=\\"0\\">
        
            <tbody><tr>
              <td style=\\"font-size:0px;word-break:break-word;\\">
                
      
    
  
      <div style=\\"height:20px;\\">
        &nbsp;
      </div>
      
    
  
    
              </td>
            </tr>
          
      </tbody></table>
    
            </td>
          </tr>
        </tbody>
      </table>
    
      </div>
    
          
              </td>
            </tr>
          </tbody>
        </table>
        
      </div>
    
      
      
    
      
      <div style=\\"border-radius: 0 0 7px 7px; border-bottom: 1px solid #f5f5f5; background: #ffffff; background-color: #ffffff; margin: 0px auto; max-width: 582px;\\" class=\\"c--email-footer c--courier-footer\\">
        
        <table style=\\"background:#ffffff;background-color:#ffffff;width:100%;\\" role=\\"presentation\\" cellspacing=\\"0\\" cellpadding=\\"0\\" border=\\"0\\" align=\\"center\\">
          <tbody>
            <tr>
              <td style=\\"border-left:1px solid #f5f5f5;border-right:1px solid #f5f5f5;direction:ltr;font-size:0px;padding:10px;padding-top:0;text-align:center;\\">
                
            
      <div style=\\"font-size:0px;text-align:left;direction:ltr;display:inline-block;vertical-align:top;width:100%;\\" class=\\"mj-column-per-100 outlook-group-fix\\">
        
      <table width=\\"100%\\" role=\\"presentation\\" cellspacing=\\"0\\" cellpadding=\\"0\\" border=\\"0\\">
        <tbody>
          <tr>
            <td style=\\"background-color:#ffffff;vertical-align:top;padding:0px;padding-bottom:10px;\\">
              
      <table width=\\"100%\\" style=\\"\\" role=\\"presentation\\" cellspacing=\\"0\\" cellpadding=\\"0\\" border=\\"0\\">
        
            <tbody><tr>
              <td style=\\"font-size:0px;padding:0px;word-break:break-word;\\" class=\\"c--text-subtext\\" align=\\"center\\">
                
      <div style=\\"font-family:Helvetica, Arial, sans-serif;font-size:11px;line-height:15px;text-align:center;color:#8F8F8F;\\">Powered By <strong><a href=\\"REPLACED\\">Courier</a></strong></div>
    
              </td>
            </tr>
          
      </tbody></table>
    
            </td>
          </tr>
        </tbody>
      </table>
    
      </div>
    
          
              </td>
            </tr>
          </tbody>
        </table>
        
      </div>
    
      
      
              </td>
            </tr>
          </tbody>
        </table>
        
      </div>
    
      
      
    
    <img src=\\"REPLACED\\">
      </div>
    
      
    
  </body></html>"
`;

exports[`send api v2 - email can get rendered subject 1`] = `"Hello, Smokey"`;
